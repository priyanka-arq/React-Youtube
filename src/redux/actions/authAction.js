import {
  getAuth,
  signOut,
  signInWithPopup,
  GoogleAuthProvider,
} from "firebase/auth";
import { firebaseApp } from "../../firebase";
import {
  LOGIN_REQUEST,
  LOGIN_SUCCESS,
  LOAD_PROFILE,
  LOGIN_FAILURE,
  LOGOUT,
} from "../actionType";

// actionCreator is  a function #login is actionCreator
// action is an Object
// login function returns dispatch fucntion
export const login = () => async (dispatch) => {
  try {
    dispatch({
      type: LOGIN_REQUEST,
    });
    const auth = getAuth();
    const provider = new GoogleAuthProvider();
    provider.addScope("https://www.googleapis.com/auth/youtube.force-ssl");
    const response = await signInWithPopup(auth, provider);

    //get accessToken
    const credential = GoogleAuthProvider.credentialFromResult(response);
    const accessToken = credential.accessToken;
    //get user profile
    const profile = {
      name: response.user.displayName,
      photoURL: response.user.photoURL,
    };

    //set accessToken and user in sessionStorage
    //stringify # profile Object
    //accessToken valid for 60min
    //#ytc is youtube-clone
    sessionStorage.setItem("ytc-access-token", accessToken);
    sessionStorage.setItem("ytc-user", JSON.stringify(profile));

    dispatch({
      type: LOGIN_SUCCESS,
      payload: accessToken,
    });

    dispatch({
      type: LOAD_PROFILE,
      payload: profile,
    });
    console.log("Auth Response", response);
  } catch (error) {
    const credential = GoogleAuthProvider.credentialFromError(error);

    dispatch({
      type: LOGIN_FAILURE,
      payload: error.message,
    });
  }
};

export const logout = () => async (dispatch) => {
  try {
    const auth = getAuth();
    await signOut(auth);

    dispatch({
      type: LOGOUT,
    });

    sessionStorage.removeItem("ytc-access-token");
    sessionStorage.removeItem("ytc-user");
  } catch (error) {
    console.log(error.message);
  }
};
