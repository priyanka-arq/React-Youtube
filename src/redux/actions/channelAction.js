import networkRequest from "../../api";
import {
  CHANNEL_DETAILS_FAILURE,
  CHANNEL_DETAILS_REQUEST,
  CHANNEL_DETAILS_SUCCESS,
} from "../actionType";

export const getChannelDetails = (id) => async (dispatch) => {
  dispatch({
    type: CHANNEL_DETAILS_REQUEST,
  });
  try {
    const response = await networkRequest.get("/channels", {
      params: {
        part: "snippet,statistics, contentDetails",
        id: id,
      },
    });

    dispatch({
      type: CHANNEL_DETAILS_SUCCESS,
      payload: response.data.items[0],
    });
  } catch (error) {
    dispatch({
      type: CHANNEL_DETAILS_FAILURE,
      payload: error.message,
    });
  }
};
