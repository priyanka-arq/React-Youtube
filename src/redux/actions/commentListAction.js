import networkRequest from "../../api";
import {
  COMMENTS_LIST_FAILURE,
  COMMENTS_LIST_SUCCESS,
  COMMENTS_LIST_REQUEST,
  CREATE_COMMENT_SUCCESS,
  CREATE_COMMENT_FAILURE,
} from "../actionType";

export const getCommentsDetails = (id) => async (dispatch) => {
  dispatch({
    type: COMMENTS_LIST_REQUEST,
  });
  try {
    const { data } = await networkRequest.get("/commentThreads", {
      params: {
        part: "snippet ",
        videoId: id,
      },
    });

    dispatch({
      type: COMMENTS_LIST_SUCCESS,
      payload: data.items,
    });
  } catch (error) {
    console.log(error.message);
    dispatch({
      type: COMMENTS_LIST_SUCCESS,
    });
  }
};

export const addComment = (id, text) => async (dispatch, getState) => {
  try {
    const obj = {
      snippet: {
        videoId: id,
        topLevelComment: {
          snippet: {
            textOriginal: text,
          },
        },
      },
    };

    await networkRequest.post("/commentThreads", obj, {
      params: {
        part: "snippet ",
      },
      headers: {
        Authorization: `Bearer ${getState().auth.accessToken}`,
      },
    });

    dispatch({
      type: CREATE_COMMENT_SUCCESS,
    });
    //feetch all commnents immmedietly after writing adding comment
    //setTimeout(() => dispatch(getCommentsDetails(id)), 3000);
  } catch (error) {
    console.log(error.message);
    dispatch({
      type: CREATE_COMMENT_FAILURE,
    });
  }
};
