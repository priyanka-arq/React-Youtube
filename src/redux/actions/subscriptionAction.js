import networkRequest from "../../api";
import {
  SET_SUBSCRIPTION_STATUS,
  SUBSCRIBE_CHANNEL_FAILURE,
  SUBSCRIBE_CHANNEL_REQUEST,
  SUBSCRIBE_CHANNEL_SUCCESS,
  SUBSCRIPTIONS_CHANNEL_FAILURE,
  SUBSCRIPTIONS_CHANNEL_REQUEST,
  SUBSCRIPTIONS_CHANNEL_SUCCESS,
} from "../actionType";

export const checkSubscriptionStatus = (id) => async (dispatch, getState) => {
  try {
    const { data } = await networkRequest.get("/subscriptions", {
      params: {
        part: "snippet",
        forChannelId: id,
        mine: true,
      },
      headers: {
        Authorization: `Bearer ${getState().auth.accessToken}`,
      },
    });

    console.log(data);
    dispatch({
      type: SET_SUBSCRIPTION_STATUS,
      payload: data.items.length !== 0,
    });
  } catch (error) {
    console.log(error.message);
  }
};

//get subscription video channel
export const getSubscribedChannels = () => async (dispatch, getState) => {
  dispatch({
    type: SUBSCRIPTIONS_CHANNEL_REQUEST,
  });
  try {
    const { data } = await networkRequest.get("/subscriptions", {
      params: {
        part: "snippet, contentDetails",
        mine: true,
        maxResults: 20,
      },
      headers: {
        Authorization: `Bearer ${getState().auth.accessToken}`,
      },
    });
    dispatch({
      type: SUBSCRIPTIONS_CHANNEL_SUCCESS,
      payload: data.items,
    });
  } catch (error) {
    dispatch({
      type: SUBSCRIPTIONS_CHANNEL_FAILURE,
      payload: error.response.data,
    });
  }
};

//Subscribe Channel
export const subascribeChannel = (channelId) => async (dispatch, getState) => {
  dispatch({
    type: SUBSCRIBE_CHANNEL_REQUEST,
  });
  try {
    const obj = {
      snippet: {
        resourceId: {
          channelId: channelId,
        },
      },
    };
    const { data } = await networkRequest.post("/subscriptions", obj, {
      params: {
        part: "snippet, contentDetails, subscriberSnippet",
      },
      headers: {
        Authorization: `Bearer ${getState().auth.accessToken}`,
      },
    });

    dispatch({
      type: SUBSCRIBE_CHANNEL_SUCCESS,
    });
  } catch (error) {
    console.log(error.message);
    dispatch({
      type: SUBSCRIBE_CHANNEL_FAILURE,
      payload: error.response.data,
    });
  }
};
