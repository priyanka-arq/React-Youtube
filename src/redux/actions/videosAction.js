import networkRequest from "../../api";
import {
  CHANNEL_VIDEOS_FAILURE,
  CHANNEL_VIDEOS_REQUEST,
  CHANNEL_VIDEOS_SUCCESS,
  HOME_VIDEOS_FAILURE,
  HOME_VIDEOS_REQUEST,
  HOME_VIDEOS_SUCCESS,
  LIKED_VIDEOS_FAILURE,
  LIKED_VIDEOS_REQUEST,
  LIKED_VIDEOS_SUCCESS,
  LIKED_VIDEO_FAILURE,
  LIKED_VIDEO_SUCCESS,
  LIKE_VIDEO_REQUEST,
  RELATED_VIDEOS_FAILURE,
  RELATED_VIDEOS_REQUEST,
  RELATED_VIDEOS_SUCCESS,
  SEARCH_VIDEO_FAILURE,
  SEARCH_VIDEO_REQUEST,
  SEARCH_VIDEO_SUCCESS,
  SELECTED_VIDEO_FAILURE,
  SELECTED_VIDEO_REQUEST,
  SELECTED_VIDEO_SUCCESS,
} from "../actionType";

export const getPopularVideos = () => async (dispatch, getState) => {
  try {
    dispatch({
      type: HOME_VIDEOS_REQUEST,
    });
    const response = await networkRequest.get("/videos", {
      params: {
        part: "snippet,contentDetails,statistics",
        chart: "mostPopular",
        regionCode: "IN",
        maxResults: 20,
        pageToken: getState().homeVideos.nextPageToken,
      },
    });

    dispatch({
      type: HOME_VIDEOS_SUCCESS,
      payload: {
        videos: response.data.items,
        nextPageToken: response.data.nextPageToken,
        category: "All",
      },
    });
  } catch (error) {
    console.log(error.message);
    dispatch({
      type: HOME_VIDEOS_FAILURE,
      payload: error.message,
    });
  }
};

//display each category on click
export const getVideosByCategory = (keyword) => async (dispatch, getState) => {
  try {
    dispatch({
      type: HOME_VIDEOS_REQUEST,
    });
    const response = await networkRequest.get("/search", {
      params: {
        part: "snippet",
        maxResults: 20,
        pageToken: getState().homeVideos.nextPageToken,
        q: keyword,
        type: "video",
      },
    });

    dispatch({
      type: HOME_VIDEOS_SUCCESS,
      payload: {
        videos: response.data.items,
        nextPageToken: response.data.nextPageToken,
        category: keyword,
        homeVideosStatus: false,
      },
    });

    console.log("response", response);
  } catch (error) {
    console.log(error.message);
    dispatch({
      type: HOME_VIDEOS_FAILURE,
      payload: error.message,
    });
  }
};

//display indiviual video on new page on click
export const getVideoById = (id) => async (dispatch) => {
  dispatch({
    type: SELECTED_VIDEO_REQUEST,
  });
  try {
    const response = await networkRequest.get("/videos", {
      params: {
        part: "snippet,statistics",
        id: id,
      },
    });

    dispatch({
      type: SELECTED_VIDEO_SUCCESS,
      payload: response.data.items[0],
    });
  } catch (error) {
    console.log(error.message);
    dispatch({
      type: SELECTED_VIDEO_FAILURE,
      payload: error.message,
    });
  }
};

//display related videos on sidebar
export const getRelatedVieos = (id) => async (dispatch) => {
  dispatch({
    type: RELATED_VIDEOS_REQUEST,
  });
  try {
    const response = await networkRequest.get("/search", {
      params: {
        part: "snippet",
        relatedToVideoId: id,
        maxResults: 15,
        type: "video",
      },
    });

    dispatch({
      type: RELATED_VIDEOS_SUCCESS,
      payload: response.data.items,
    });
  } catch (error) {
    console.log(error.message);
    dispatch({
      type: RELATED_VIDEOS_FAILURE,
    });
  }
};

//Search bar actions
export const getVideosBySearch = (keyword) => async (dispatch, getState) => {
  try {
    dispatch({
      type: SEARCH_VIDEO_REQUEST,
    });
    const response = await networkRequest.get("/search", {
      params: {
        part: "snippet",
        maxResults: 20,
        pageToken: getState().homeVideos.nextPageToken,
        q: keyword,
        type: "video, channel",
      },
    });

    dispatch({
      type: SEARCH_VIDEO_SUCCESS,
      payload: response.data.items,
    });

    console.log("response", response);
  } catch (error) {
    console.log(error.message);
    dispatch({
      type: SEARCH_VIDEO_FAILURE,
      payload: error.message,
    });
  }
};

//display videos by channel
export const getChannelVideos = (id) => async (dispatch) => {
  dispatch({
    type: CHANNEL_VIDEOS_REQUEST,
  });
  try {
    //1.get upload playlist id
    const {
      data: { items },
    } = await networkRequest.get("/channels", {
      params: {
        part: "contentDetails",
        id: id,
      },
    });

    console.log("channel video response", items);
    //get uploadPlayListId from channel response and use it to get all videos os that channel
    const uploadPlayListId = items[0].contentDetails.relatedPlaylists.uploads;

    //1.get the videos using uploadPlayListId that received form 1st request
    const { data } = await networkRequest.get("/playlistItems", {
      params: {
        part: "contentDetails, snippet",
        playlistId: uploadPlayListId,
        maxResults: 30,
      },
    });

    dispatch({
      type: CHANNEL_VIDEOS_SUCCESS,
      payload: data.items,
    });
  } catch (error) {
    console.log(error.message);
    dispatch({
      type: CHANNEL_VIDEOS_FAILURE,
      payload: error.message,
    });
  }
};

//Liked videos
export const getLikedVideos = (keyword) => async (dispatch, getState) => {
  try {
    dispatch({
      type: LIKED_VIDEOS_REQUEST,
    });
    const response = await networkRequest.get("/videos", {
      params: {
        part: "snippet, contentDetails,  statistics",
        myRating: "like",
      },
      headers: {
        Authorization: `Bearer ${getState().auth.accessToken}`,
      },
    });

    console.log("liked videos", response.data.items);
    dispatch({
      type: LIKED_VIDEOS_SUCCESS,
      payload: response.data.items,
    });

    console.log("response", response);
  } catch (error) {
    console.log(error.message);
    dispatch({
      type: LIKED_VIDEOS_FAILURE,
      payload: error.message,
    });
  }
};

//Liked videos
export const likeVideo = (id) => async (dispatch, getState) => {
  try {
    dispatch({
      type: LIKE_VIDEO_REQUEST,
    });
    const response = await networkRequest.post("/videos/rate", {
      params: {
        part: "snippet, contentDetails,  statistics",
        rating: "like",
      },
      headers: {
        Authorization: `Bearer ${getState().auth.accessToken}`,
      },
    });

    dispatch({
      type: LIKED_VIDEO_SUCCESS,
    });

    console.log("response", response);
  } catch (error) {
    console.log(error.message);
    dispatch({
      type: LIKED_VIDEO_FAILURE,
    });
  }
};
