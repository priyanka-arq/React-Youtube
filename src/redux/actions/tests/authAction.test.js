import { login, logout } from "../../actions/authAction";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import {
  LOGIN_REQUEST,
  LOGIN_SUCCESS,
  LOAD_PROFILE,
  LOGIN_FAILURE,
  LOGOUT,
} from "../../actionType";

const createMockStore = configureMockStore([thunk]);

let profile;
let accessToken;
let error;
let actions;
let action;

beforeEach(() => {
  profile = {
    name: "Priyanka Patel",
    photoURL:
      "https://lh3.googleusercontent.com/a-/AOh14GgEX3SB8FVF-3OLqmdT3zAW0vRL6alhlseVN0bn=s96-c",
  };
  (error =
    "Firebase: Error (auth/operation-not-supported-in-this-environment)."),
    (accessToken = "AIzaSyCLg4cHF4d0Sj_UwQQhV1JPvCngteYtNkM");
});

describe("auth action", () => {
  test("LOGIN_SUCCESS action", () => {
    action = { type: LOGIN_SUCCESS, payload: accessToken };
    expect(action).toEqual({
      type: LOGIN_SUCCESS,
      payload: accessToken,
    });
  });

  test("login action", () => {
    const expected = [
      { type: LOGIN_REQUEST },
      {
        type: LOGIN_SUCCESS,
        payload: accessToken,
      },
      {
        type: LOAD_PROFILE,
        payload: profile,
      },
      {
        type: LOGIN_FAILURE,
        payload: error,
      },
    ];
    const store = createMockStore();
    store.dispatch(login()).then(() => {
      actions = store.getActions();
      expect(store.getActions()).toMatchSnapshot();
      expect(actions[0]).toEqual({
        type: LOGIN_REQUEST,
      });
      expect(actions[1]).toEqual({
        type: LOGIN_FAILURE,
        payload: error,
      });
    });
  });
});

describe("LOGOUT", () => {
  const store = createMockStore();
  test("logout action", () => {
    store.dispatch(logout()).then(() => {
      actions = store.getActions();
      expect(store.getActions()).toMatchSnapshot();
      expect(actions[0]).toEqual({
        type: LOGOUT,
      });
    });
  });
});
