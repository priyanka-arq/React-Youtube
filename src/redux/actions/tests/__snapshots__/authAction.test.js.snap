// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LOGOUT logout action 1`] = `
Array [
  Object {
    "type": "LOGOUT",
  },
]
`;

exports[`auth action login action 1`] = `
Array [
  Object {
    "type": "LOGIN_REQUEST",
  },
  Object {
    "payload": "Firebase: Error (auth/operation-not-supported-in-this-environment).",
    "type": "LOGIN_FAILURE",
  },
]
`;
